Sources/Basic/ArgumentParser.swift

```diff
@@ -0,0 +1,223 @@
+                }
+                result = intValue
+            case .string:
+                result = value
+            case .stringArray:
+                var array = [String]()
+                // Get the previously added results if present.
+                if let previousResult = results[argument] as? [String] {
+                    array = previousResult
+                }
+                array.append(value)
+                result = array
+            }
+            // Save the result.
+            results[argument] = result
+        }
+
+        /// Get an option argument's value from the results.
+        ///
+        /// Since the options are optional, their result may or may not
+        /// be present.
+        public func get<T>(_ arg: OptionArgument) -> T? {
+            return results[arg] as? T
+        }
+
+        /// Get a positional argument's value.
+        public func get<T>(_ arg: PositionalArgument) -> T {
+            return results[arg] as! T
+        }
+    }
+
+    /// List of arguments added to this parser.
+    private var arguments = [CommandLineArgument]()
+
+    public init() {
+    }
+
+    /// Adds an option to the parser.
+    public mutating func add(option: String, kind: CommandLineArgument.Kind, usage: String? = nil) -> OptionArgument {
+        let arg = OptionArgument(name: option, kind: kind, usage: usage)
+        arguments.append(arg)
+        return arg
+    }
+
+    /// Adds an argument to the parser.
+    public mutating func add(argument: String, kind: CommandLineArgument.Kind, usage: String? = nil) -> PositionalArgument {
+        let arg = PositionalArgument(name: argument, kind: kind, usage: usage)
+        arguments.append(arg)
+        return arg
+    }
+
+    /// Parses the provided array and return the result.
+    public mutating func parse(_ args: [String] = []) throws -> ArgumentParserResult {
+        var result = ArgumentParserResult()
+
+        // Create options map to quickly look up the arguments.
+        let optionsMap = Dictionary(items: arguments.flatMap { $0 as? OptionArgument }.map { ($0.name, $0) })
+        // Seperate out positional arguments.
+        let positionalArgs = arguments.flatMap { $0 as? PositionalArgument }
+        // Create iterators.
+        var positionalArgsIterator = positionalArgs.makeIterator()
+        var it = args.makeIterator()
+
+        while let arg = it.next() {
+            if arg.isPositionalArg {
+                guard let positionalArg = positionalArgsIterator.next() else {
+                    throw ArgumentParserError.unexpectedArgument(arg)
+                }
+                try result.addResult(for: positionalArg, value: arg)
+            } else {
+                guard let option = optionsMap[arg] else {
+                    throw ArgumentParserError.unknown(option: arg)
+                }
+                // FIXME: Find a better way to handle bool which doesn't have associated values.
```

Comment:
I would handle this by using the result object to map from the argument to an object provided by the "argument kind" protocol -- that protocol would also have a method to extract the value from that object when clients want to get the "typed" value (I think, haven't worked it through in my head).